### ISSUE TITLE
twisted.application.reactors.installReactor should return the installed reactor.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6596]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#6596|
|Type|enhancement|
|Created|2013-06-25 18:14:19Z|
|Branch|https://github.com/twisted/twisted/tree/install-reactor-6596|

In general, it is preferred to pass a reactor around explicitly, rather than just importing it from `twisted.internet`. However, when testing code that explicitly creates a reactor, the only way for that code to get a hold of the just installed reactor is to import it.

This means that that to test code that explicitly installs the reactor and then uses it, you need to stub out the code that installs the reactor *and* the global reactor.

If instead, the reactor that is installed is returned, then only the former needs to be stubbed out.

<details><summary>Searchable metadata</summary>

```
trac-id__6596 6596
type__enhancement enhancement
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__branches_install_reactor_6596 branches/install-reactor-6596
branch_author__tomprince tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1372184059000000 1372184059000000
changetime__1373783385000000 1373783385000000
version__None None
owner__tom_prince tom.prince
cc__lukmdo
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_1"></a>|@tomprince commented|
|-|-|

(In [38916]) Branching to install-reactor-6596.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_2"></a>|@tomprince commented|
|-|-|

(In [38917]) Fix existing test, to match its docstring.

`test_reactorInstallation` claims to test that it test `Reactor.install`,
but in fact only tests a stubbed out version.

This changes the test to actually test it, by modifying `sys.modules`.

Refs: [#6596](https://github.com/twisted/twisted/issues/6596)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince removed owner|
|-|-|

I'd be inclined to change `twisted.application.reactors.Reactor` to return the reactor, but that wouldn't be compatible with existing plugins that don't use that to implement `IReactorInstaller`.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince commented|
|-|-|

(In [38922]) A bit more coverage.

Refs: [#6596](https://github.com/twisted/twisted/issues/6596)

**Comment by twisted-trac** [comment]
|<img alt="lukmdo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|lukmdo commented|
|-|-|

*  _IReactorInstaller.install_ could need some extra docstring to  reflect _from twisted.internet import reactor; return reactor_ (it is tested using _patch_)

*  I am twisted newcomer not sure if letting _IReactorInstaller.install_ return _None_ || _reactorObject_ would not make more sense then basing/modifying namespace

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @tomprince|
|-|-|

Thanks for your work on this tom.prince and lukmdo.

In addition to the above:

  1. `test_installReactorReturnsReactor` uses `assertEqual`.  I think this is an appropriate place to use `assertIdentical`.  The API should return exactly the expected object, by identity, not merely something semantically equivalent to that object.  I guess?  Not that there are a lot of things that are semantically equivalent to any particular reactor.
  1. In the news fragment, "just-installed" is perhaps slightly more correct than "just installed".

I agree with your compatibility concerns regarding `Reactor.install`.  If it were changed, then user code could only take advantage of that if they also handled all other implementations that don't return the reactor.  So leaving it alone seems like the right thing to do for now.

Thanks again.  Please merge after addressing these points to your satisfaction.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince commented|
|-|-|

Replying to [lukmdo](#note_5):
> I am twisted newcomer not sure if letting `IReactorInstaller.install` return `None` || `reactorObject` would not make more sense then basing/modifying namespace

In an ideal world, if the api was being designed now, `IReactorInstaller.install` would always return the reactor is just installed. But there exist[1] external implementations of `IReactorInstaller.install` that don't do that. Thus we can't depend on that behavior, so we would have to handle that case[2] in `installReactor` anyway.

[1] Even if there were no publicly availble external implementations, our [CompatibilityPolicy](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy) requires we behave as if there were.

[2] In fact, we can't even assume that the return value will be a reactor or `None`. Since `IReactorInstaller.install` doesn't document a return value, it can potentially return anything.

> `IReactorInstaller.install`` could need some extra docstring to reflect `from twisted.internet import reactor; return reactor`

I just did some digging, and I couldn't find any documentation that explains what "installing a reactor" means. I filed some tickets related to this ([#6623](https://github.com/twisted/twisted/issues/6623), [#6624](https://github.com/twisted/twisted/issues/6624), [#6625](https://github.com/twisted/twisted/issues/6625)).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince commented|
|-|-|

(In [39158]) Use assertIdentical instead of assertEqual to compare reactors.

Refs: [#6596](https://github.com/twisted/twisted/issues/6596)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince commented|
|-|-|

(In [39159]) topfile punctuation

Refs: [#6596](https://github.com/twisted/twisted/issues/6596)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince set status to `closed`|
|-|-|

(In [39160]) Merge install-reactor-6596: Make twisted.application.reactors.installReactor return the installed reactor.

Author: tom.prince
Reviewers: lukmdo, exarkun
Fixes: [#6596](https://github.com/twisted/twisted/issues/6596)

This allows code that installs a reactor by name to be tested by
just stubbing out `installReactor`, rather than both it *and* the
global reactor.
